### PR TITLE
FOPTS-212 - Cloud Cost Anomaly Alerts - Filter on dimensions

### PR DESCRIPTION
**Pull request by vadgun**
### Description

Policy now receive a key-value pair for filter dimension from API Endpoint

### Issues Resolved

https://flexera.atlassian.net/browse/FOPTS-212

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### Links

Applied policy Flexera one demo 
https://app.flexera.com/orgs/28010/automation/applied-policies/projects/123559?policyId=642af0a394f0030001547e3b

image

### CODE REVIEW

File: cost/cloud_cost_anomaly_alerts/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,14 @@
 # Changelog
 
+## v2.5
+
+- added a filter for dimensions when user wants to filter by key pair value like Service=AmazonEC2
+
+## v2.4
```
Comment (juanestrahcl) [inline cost/cloud_cost_anomaly_alerts/CHANGELOG.md:7]: What is the reason to update version twice this PR?

File: cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt [hunk 2]
```diff
@@ -429,10 +530,41 @@ end
 ###############################################################################
 
 policy "budget_alert" do
+
```
Comment (juanestrahcl) [inline cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt:531]: Remove this blank line.

File: cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt [hunk 3]
```diff
@@ -429,10 +528,41 @@ end
...
+    escalate $esc_budget_alert
+  end
+
+
   validate $ds_filter_anomalies do
     summary_template "Cloud Cost Anomalies: {{ len data.anomalies}} rows of detected anomalies for dimensions: {{ data.dimensionsReadable }}"
     detail_template <<-EOS
+
 Cost anomalies reference link: {{data.message}}
+\n
```
Comment (juanestrahcl) [inline cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt:553]: Fix indentation of this block of code.

**Review by juanestrahcl — COMMENTED** [review]
Aside of comments I think you should go forward and start using new tickets' names, I think this is "FOPTS-212", also there are things that you can set in your PR like:

- Assign it to yourself.
- Add reviewers.

File: cost/cloud_cost_anomaly_alerts/CHANGELOG.md (hunk 1 above)
Comment (vadgun) [reply cost/cloud_cost_anomaly_alerts/CHANGELOG.md:7]: There no reason, the criterion was taken from the ambiguity that exists in the 2 tickets (HCLIND-2 & HCLIND-59). Either make them an update itself or separately.

Comment (vadgun) [reply cost/cloud_cost_anomaly_alerts/CHANGELOG.md:3]: i left only one update version, thank you

**Review by vadgun — COMMENTED** [review]
Good

**Review by juanestrahcl — DISMISSED** [review]

File: cost/cloud_cost_anomaly_alerts/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## v2.4
+
+- added a filter for dimensions when user wants to filter by key pair value like Service=AmazonEC2
```
Comment (degarajesh) [inline cost/cloud_cost_anomaly_alerts/CHANGELOG.md:5]: Start with capital letter, Added...

Comment (degarajesh) [reply cost/cloud_cost_anomaly_alerts/CHANGELOG.md:5]: How about ?

Added a filter for dimensions where the user wishes to filter by key pair values, such as Service=AmazonEC2

File: cost/cloud_cost_anomaly_alerts/CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,11 @@
...
+- added an input list for a mix of dimensions
```
Comment (degarajesh) [inline cost/cloud_cost_anomaly_alerts/CHANGELOG.md:6]: Please start with Capital Letter, Added.

File: cost/cloud_cost_anomaly_alerts/CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,11 @@
...
+- validated each input from the user input list, if all are valid dimensions, it will run an anomaly report. Otherwise, when the user inputs at least one invalid dimension, policy will fail, escalating to emailing you about invalid dimension you used.
```
Comment (degarajesh) [inline cost/cloud_cost_anomaly_alerts/CHANGELOG.md:7]: Can we change it to ?

If at least one invalid dimension is entered by the user, the policy will fail, resulting in an email to you about the invalid dimension.

File: cost/cloud_cost_anomaly_alerts/README.md [hunk 7]
```diff
@@ -21,8 +21,10 @@ Any point outside of the bands is considered as anomalous. If multiple cost anom
 - *Minimum Period Spend* - minimum daily spend needed to record an anomaly as an incident.
   - Minimum Value: 0
   - Default: 1000
-- *Cost Anomaly Dimensions* - Dimensions to group data for the analysis. Multiple dimensions can be provided. Currently, custom tags cannot be used as dimensions.
+- *Cost Anomaly Dimensions* - Dimensions to group data for the analysis. Multiple dimensions can be provided.
   - Default: ["Cloud Vendor Account", "Cloud Vendor", "Service"]
```
Comment (degarajesh) [inline cost/cloud_cost_anomaly_alerts/README.md:25]: Dimensions to group data for analysis. It is possible to provide multiple dimensions.

Can we change ?

File: cost/cloud_cost_anomaly_alerts/README.md [hunk 8]
```diff
@@ -21,8 +21,10 @@ Any point outside of the bands is considered as anomalous. If multiple cost anom
...
+  - Id(s) or name(s) for Rule Based Dimensions to group data for the analysis. Multiple rbd id(s) or name(s) can be provided. This allows for better visualization of data that was not tagged or difficult to tag in the native cloud platform, these new dimensions in Optima will be populated according to rules you define, you can use the new dimensions to look at your cost from new meaningful perspectives, a common use case is to create several dimensions like 'CCO Environments' or 'Product Line', populated based on know account values.
+  - Filters can now be added to the same values ​​of the dimensions, you can filter 1 to 1 value per dimension e.g. Service=AmazonEC2 , Cloud Vendor=AWS , Cloud Vendor Account Name=Flexera-prod.
```
Comment (degarajesh) [inline cost/cloud_cost_anomaly_alerts/README.md:27]: Can we change to 

ID(s) or name(s) for Rule Based Dimensions to group data for analysis. Multiple RBD ID(s) or name(s) can be provided. This allows for better visualization of data that was not tagged or difficult to tag in the native cloud platform, these new dimensions in Optima will be populated according to rules you define, you can use the new dimensions to look at your cost from new meaningful perspectives, a common use case is to create several dimensions like 'CCO Environments' or 'Product Line', populated based on known account values. - Filters can now be added to the same values ​​of the dimensions. You can filter 1 to 1 value per dimension e.g. Service=AmazonEC2 , Cloud Vendor=AWS , Cloud Vendor Account Name=Flexera-prod.

File: cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt [hunk 9]
```diff
@@ -39,16 +39,14 @@ parameter "param_cost_metric" do
...
   description "Select the cost metric for your report.  See the README file for more details"
 end
 
-##
 parameter "param_dimensions" do
   type "list"
   label "Cost Anomaly Dimensions"
-  description "Dimensions to run anomaly report. Unable to take in custom tags as a dimension"
-  allowed_values "Cloud Vendor","Cloud Vendor Account Name","Cloud Vendor Account","Manufacturer Name","Spend Type","Flexera Capability","Category","Instance Type","Region","Operation","Resource Type","Service","Billing Account ID","Usage Unit","Usage Type","Line Item Type","Purchase Option","Resource Group","Bill Entity","Bill Source","Invoice ID"
+  description "Add please the exact case sensitive name(s) or id(s) of Default or Rule Based Dimension you wish to run anomaly report. e.g. 'CCO Environments' for name / e.g. 'rbd_cost_center' for id. Also you can add more filters for anomalies on specific dimension by adding 'Vendor=AWS'; first criteria will be mandatory for consequent filters"
```
Comment (degarajesh) [inline cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt:49]: Can we change to ?
Add please the exact case sensitive name(s) or id(s) of Default or Rule Based Dimension you wish to run anomaly report. e.g. 'CCO Environments' for name / e.g. 'rbd_cost_center' for id. Also you can add more filters for anomalies on specific dimension by adding 'Vendor=AWS'; first criteria will be mandatory for consequent filters

File: cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt [hunk 10]
```diff
@@ -267,38 +309,115 @@ script "js_get_anomalies", type: "javascript" do
...
+        expression.value = filter.value;
+        expressions.push(expression)
+      })
+
+      var request = {
+        auth: "auth_flexera",
+        verb: "POST",
+        host: rs_optima_host,
+        path: "/bill-analysis/orgs/" + org_id + "/anomalies/report",
+        body_fields: {
```
Comment (degarajesh) [inline cost/cloud_cost_anomaly_alerts/cloud_cost_anomaly_alerts.pt:371]: Can we optimize this Code.

**Review by juanestrahcl — APPROVED** [review]

**Review by bryankaraffa — COMMENTED** [review]
Please resolve merge conflicts and we can re-review or approve and merge if no other changes needed.  Thank you!

**Comment by degarajesh** [comment]
> Please resolve merge conflicts and we can re-review or approve and merge if no other changes needed. Thank you!

@bryankaraffa

**Pull request merged**